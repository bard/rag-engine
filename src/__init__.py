from . import ingest
from . import query
