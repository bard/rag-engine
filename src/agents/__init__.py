from .document_processor import DocumentProcessor
from .vector_store import VectorStore

__all__ = ["DocumentProcessor", "VectorStore"]
