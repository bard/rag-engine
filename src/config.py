from langchain_core.runnables.config import RunnableConfig
from pydantic import BaseModel, Field
from typing import Self, Union, Literal


class PineconeVectorStoreBackend(BaseModel):
    type: Literal["pinecone"]
    index_name: str
    api_key: str
    embeddings_type: Literal["openai"]
    score_threshold: float = 0.1


class MockVectorStoreBackend(BaseModel):
    type: Literal["mock"]
    score_threshold: float = 0.1


class ChromaVectorStoreBackend(BaseModel):
    type: Literal["chroma"]
    path: str
    collection_name: str
    embeddings_type: Literal["local-minilm", "openai"]
    score_threshold: float = 0.1


type VectorStoreBackend = Union[
    PineconeVectorStoreBackend, MockVectorStoreBackend, ChromaVectorStoreBackend
]


class OpenaiLlmBackend(BaseModel):
    type: Literal["openai"]
    model: str
    api_key: str


class AnthropicLlmBackend(BaseModel):
    type: Literal["anthropic"]
    model: str
    api_key: str


class DbBackend(BaseModel):
    url: str


type LlmBackend = Union[OpenaiLlmBackend, AnthropicLlmBackend]


class AgentConfig(BaseModel):
    db: DbBackend
    llm: LlmBackend = Field(discriminator="type")
    vector_store: VectorStoreBackend = Field(discriminator="type")

    @classmethod
    def from_runnable_config(cls, config: RunnableConfig) -> Self:
        """Validate runnable config"""
        configurable = config.get("configurable")
        assert configurable is not None
        return cls(**configurable)
