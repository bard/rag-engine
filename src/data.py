from langchain.schema import Document
from typing import Dict, Any, Self
from pydantic import BaseModel
from bs4 import BeautifulSoup
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase


class InsuranceRecord(BaseModel):
    """Pydantic model for insurance record data"""

    year: int
    average_expenditure: float
    percent_change: float
    source_content: str

    def to_langchain_document(self, source_url: str) -> Document:
        """
        Convert this InsuranceRecord to a LangChain Document.

        Args:
            source_url: The URL where the record was sourced from

        Returns:
            Document object containing the record data and metadata
        """
        content = "\n".join(
            [
                f"Year: {self.year}, "
                f"Expenditure: ${self.average_expenditure:.2f}, "
                f"Change: {self.percent_change}%"
            ]
        )

        return Document(
            page_content=content,
            id=f"insurance-record-{self.year}",
            metadata={
                "source_url": source_url,
                "source_content": self.source_content,
                "year": self.year,
                "average_expenditure": self.average_expenditure,
                "percent_change": self.percent_change,
            },
        )

    @classmethod
    def from_html_content(cls, html_content: str) -> list[Self]:
        """
        Parse the auto insurance expenditure table and return a list of dictionaries.
        Each dictionary represents a row with year, expenditure, and percent change.
        """
        # Create BeautifulSoup object
        soup = BeautifulSoup(html_content, "html.parser")

        # Find the innermost table with the actual data
        tables = soup.find_all("table")
        data_table = tables[2]  # The third table contains our data

        # Initialize list to store results
        results = []

        # Process each row in tbody
        for row in data_table.tbody.find_all("tr"):
            cols = row.find_all("td")

            # Extract values
            year = int(cols[0].text.strip())

            # Clean and convert expenditure to float
            expenditure = cols[1].text.strip()
            expenditure = expenditure.replace("$", "").replace(",", "")
            expenditure = float(expenditure)

            # Clean and convert percent change to float
            percent_change = cols[2].text.strip()
            percent_change = percent_change.replace("%", "")
            percent_change = float(percent_change)

            # Create dictionary for current row
            record = cls(
                year=year,
                average_expenditure=expenditure,
                percent_change=percent_change,
                source_content=str(row),
            )
            results.append(record)

        return results


class SqlAlchemyBase(DeclarativeBase):
    pass


class SQLInsuranceRecord(SqlAlchemyBase):
    """SQLAlchemy model for insurance record data"""

    __tablename__ = "insurance_records"

    id: Mapped[int] = mapped_column(primary_key=True)
    year: Mapped[int] = mapped_column(nullable=False)
    average_expenditure: Mapped[float] = mapped_column(nullable=False)
    percent_change: Mapped[float] = mapped_column(nullable=False)
    source_content: Mapped[str] = mapped_column(nullable=False)

    def __repr__(self) -> str:
        return (
            f"SQLInsuranceRecord(id={self.id}, "
            f"year={self.year}, "
            f"average_expenditure={self.average_expenditure:.2f}, "
            f"percent_change={self.percent_change})"
        )


class RawGenericTabularData(BaseModel):
    title: str
    data: list[Dict[str, Any]]


class GenericTabularData(RawGenericTabularData):
    content_hash: str

    def to_langchain_document(self) -> Document:
        """
        Convert this GenericTabularData to a LangChain Document.

        Returns:
            Document object containing the data and metadata
        """
        content = f"Title: {self.title}\n"
        for item in self.data:
            content += "\n".join(f"{k}: {v}" for k, v in item.items())
            content += "\n\n"

        return Document(
            page_content=content,
            id=f"generic-data-{self.content_hash[:8]}",
            metadata={
                "title": self.title,
                "content_hash": self.content_hash,
                "raw_data": self.data,
            },
        )
