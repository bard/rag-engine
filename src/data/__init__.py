from .base import IndexableData
from .generic_tabular import GenericTabularData
from .insurance_average_expenditure import InsuranceAverageExpenditureData
from .textual import TextualData

__all__ = [
    "IndexableData",
    "GenericTabularData",
    "InsuranceAverageExpenditureData",
    "TextualData",
]
