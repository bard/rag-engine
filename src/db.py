import uuid
from typing import List
from datetime import datetime, timezone
from sqlalchemy.orm import Mapped, mapped_column, relationship, DeclarativeBase
from sqlalchemy import inspect, DateTime, ForeignKey


class SqlAlchemyBase(DeclarativeBase):
    pass


class SqlTopic(SqlAlchemyBase):
    """SqlAlchemy model for stored topic"""

    __tablename__ = "topics"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: str(uuid.uuid4()))
    name: Mapped[str] = mapped_column(nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        default=lambda: datetime.now(timezone.utc), nullable=False
    )

    # Relationship to KnowledgeBase
    knowledge_documents: Mapped[List["SqlKnowledgeBaseDocument"]] = relationship(
        back_populates="topic", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"SqlTopic(id={self.id}, name={self.name})"


class SqlKnowledgeBaseDocument(SqlAlchemyBase):
    """SqlAlchemy model for stored document"""

    __tablename__ = "documents"

    id: Mapped[str] = mapped_column(primary_key=True, default=lambda: str(uuid.uuid4()))
    data: Mapped[str] = mapped_column(nullable=False)
    content: Mapped[str] = mapped_column(nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        nullable=False,
    )
    topic_id: Mapped[str] = mapped_column(
        ForeignKey("topics.id"),
        # allow NULL to signify "default topic"
        nullable=True,
    )
    topic: Mapped[SqlTopic] = relationship(back_populates="knowledge_documents")

    def __repr__(self) -> str:
        return f"SqlKnowledgeBaseDocument(id={self.id}, data={self.data}, content={self.content}"

    def to_dict(self):
        return {c.key: getattr(self, c.key) for c in inspect(self).mapper.column_attrs}
