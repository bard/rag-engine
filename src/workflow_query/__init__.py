from .state import AgentState
from .graph import get_graph
from .node_classify_query import classify_query
from .node_fetch_weather_info import fetch_weather_info
from .node_generate import generate
from .node_retrieve import retrieve
from .node_rerank__STUB import rerank
from .node_reformulate_query__STUB import reformulate_query

__all__ = [
    "AgentState",
    "get_graph",
    "classify_query",
    "fetch_weather_info",
    "generate",
    "reformulate_query",
    "retrieve",
    "rerank",
]
