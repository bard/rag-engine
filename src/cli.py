import click
import logging
from os import path
from langchain_core.messages import HumanMessage, AIMessage
from sqlalchemy.orm import Session

from src import services, workflow_query, workflow_ingest, db
from src.config import Config
from src.db import SqlTopic


CONFIG = Config.from_env()


@click.group()
@click.option(
    "--log",
    help="Set logging level",
    default="INFO",
    type=click.Choice(
        ["DEBUG", "INFO", "WARNING", "ERROR", "CRITICAL"], case_sensitive=False
    ),
)
def cli(log):
    logging.basicConfig(level=getattr(logging, log.upper()))


@click.command(name="create_topic")
@click.option("--id", required=False, help="Optional topic ID")
@click.option("--name", required=True, help="Topic name")
def cmd_create_topic(id: str | None, name: str):
    db = services.get_db(CONFIG)
    with Session(db) as session:
        topic = SqlTopic(id=id, name=name)
        session.add(topic)
        session.commit()
        session.refresh(topic)

    click.echo()
    click.echo(f"Created topic '{name}' with ID: {topic.id}")
    click.echo()


@click.command(name="ingest")
@click.argument("path_or_url")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_ingest(path_or_url: str, topic_id: str | None):
    if path_or_url.startswith(("http://", "https://", "file://", "data://")):
        input_url = path_or_url
    else:
        abs_path = path.abspath(path_or_url)
        input_url = f"file://{abs_path}"
    initial_agent_state = workflow_ingest.AgentState(
        url=input_url, source_content=None, extracted_data=[], topic_id=topic_id
    )

    workflow_ingest.get_graph().invoke(initial_agent_state, CONFIG.to_runnable_config())

    click.echo()
    click.echo("Data ingested successfully")
    click.echo()


@click.command(name="list_topics")
def cmd_list_topics():
    """List all available topics."""
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topics = session.query(SqlTopic).order_by(SqlTopic.name).all()

    click.echo()
    if topics:
        click.echo("Available topics:")
        for topic in topics:
            click.echo(f"  {topic.id}: {topic.name}")
    else:
        click.echo("No topics found")
    click.echo()


@click.command(name="info")
def cmd_info():
    """Display information about the database and vector store."""
    # Get database stats
    db_engine = services.get_db(CONFIG)
    with Session(db_engine) as session:
        topic_count = session.query(SqlTopic).count()
        doc_count = session.query(db.SqlKnowledgeBaseDocument).count()

    # Get sanitized DB URL (remove credentials)
    db_url = str(CONFIG.db.url)
    if "@" in db_url:
        db_url = db_url.split("@")[1]
        db_url = f"***:***@{db_url}"

    click.echo()
    click.echo("Database Information:")
    click.echo(f"  URL: {db_url}")
    click.echo(f"  Topics: {topic_count}")
    click.echo(f"  Documents: {doc_count}")
    click.echo()


@click.command(name="initdb")
def cmd_initdb():
    engine = services.get_db(CONFIG)

    db.SqlAlchemyBase.metadata.create_all(engine)

    click.echo()
    click.echo("Database initialized successfully")
    click.echo()


@click.command(name="query")
@click.argument("user_query")
@click.option("--topic_id", required=False, help="Optional topic ID")
def cmd_query(user_query: str, topic_id: str | None):
    initial_agent_state = workflow_query.AgentState(
        messages=[HumanMessage(content=user_query)],
        retrieved_knowledge=[],
        query=None,
        topic_id=topic_id,
        external_knowledge_sources=[],
    )

    response = workflow_query.get_graph().invoke(
        initial_agent_state, CONFIG.to_runnable_config()
    )

    ai_messages = [msg for msg in response["messages"] if isinstance(msg, AIMessage)]
    click.echo()
    if ai_messages:
        click.echo(ai_messages[-1].content)
    else:
        click.echo("No AI response found")
    click.echo()


if __name__ == "__main__":
    cli.add_command(cmd_create_topic)
    cli.add_command(cmd_ingest)
    cli.add_command(cmd_info)
    cli.add_command(cmd_initdb)
    cli.add_command(cmd_list_topics)
    cli.add_command(cmd_query)
    cli()
