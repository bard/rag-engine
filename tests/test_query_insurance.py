import pprint
import pytest
from langchain.schema import HumanMessage
from langchain_core.messages import HumanMessage

from src import services
from src.workflow_query import (
    AgentState,
    get_graph,
    retrieve_from_knowledge_base,
)


def test_retrieve_with_insurance_knowledge_base(config):
    vector_store = services.get_vector_store(config)
    vector_store.add_texts(
        [
            """# Average expenditures 2012-2013
Year: 2012, Expenditure: $812.40, Change: 2.2%
Year: 2013, Expenditure: $841.06, Change: 3.5%
""",
            """# Average expenditures 2014-2015
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
""",
            """# Average expenditures 2015-2016
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
""",
        ],
        metadatas=[
            {"topic_id": "UNCATEGORIZED"},
            {"topic_id": "UNCATEGORIZED"},
            {"topic_id": "UNCATEGORIZED"},
        ],
    )
    agent_state = AgentState(
        messages=[
            HumanMessage(content="what is the expenditure in 2014?"),
        ],
        retrieved_knowledge=[],
        query="what is the expenditure in 2014?",
        topic_id=None,
        external_knowledge_sources=[],
    )

    state_update = retrieve_from_knowledge_base(
        agent_state, config.to_runnable_config()
    )

    documents_content = "\n".join(
        [doc.page_content for doc in state_update["retrieved_knowledge"]]
    )

    assert (
        documents_content
        == """# Average expenditures 2014-2015
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%

# Average expenditures 2015-2016
Year: 2014, Expenditure: $869.47, Change: 3.4%
Year: 2015, Expenditure: $896.66, Change: 3.1%
"""
    )


@pytest.mark.vcr
@pytest.mark.parametrize(
    "user_query",
    [
        "What is the trend in auto insurance costs between 2012 and 2015?",
        # "Compare insurance costs between different regions",
        # "What factors most influence insurance costs?",
        # "Generate a summary of key findings from the data",
    ],
)
def test_graph_with_insurance_queries(
    config, user_query, insurance_data_documents, snapshot
):
    services.get_vector_store(config).add_documents(insurance_data_documents)

    agent_state = AgentState(
        messages=[HumanMessage(content=user_query)],
        retrieved_knowledge=[],
        query=None,
        topic_id=None,
        external_knowledge_sources=[],
    )

    response = get_graph().invoke(agent_state, config.to_runnable_config())

    assert response.get("messages", [])[-1].content == snapshot
