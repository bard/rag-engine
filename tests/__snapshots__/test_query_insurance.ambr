# serializer version: 1
# name: test_graph_with_insurance_queries[What is the trend in auto insurance costs between 2012 and 2015?]
  'Between 2012 and 2015, the trend in auto insurance costs was an increase. The average expenditure rose from $812.40 in 2012 to $896.66 in 2015, with annual percentage increases of 2.2% in 2012, 3.5% in 2013, 3.4% in 2014, and 3.1% in 2015.'
# ---
